src/anet.c

```diff
@@ -130,57 +130,135 @@ int anetCloexec(int fd) {
+     * you can use the TCP_KEEPALIVE_ABORT_THRESHOLD socket option to change the time threshold for aborting a TCP connection.
+     * The option value is an unsigned integer in milliseconds. The value zero indicates that TCP should never time out and 
+     * abort the connection when probing. The system default is controlled by the TCP ndd parameter tcp_keepalive_abort_interval. 
+     * The default is eight minutes.
+
+     * - The second implementation is activated if socket option TCP_KEEPINTVL and/or TCP_KEEPCNT are set. 
+     * The time between each consequent probes is set by TCP_KEEPINTVL in seconds. 
+     * The minimum value is ten seconds. The maximum is ten days, while the default is two hours. 
+     * The TCP connection will be aborted after certain amount of probes, which is set by TCP_KEEPCNT, without receiving response.
+     */
+
+    idle = interval;
+    if (idle < 10) idle = 10; // kernel expects at least 10 seconds
+    if (idle > 10*24*60*60) idle = 10*24*60*60; // kernel expects at most 10 days
+    
+    /* `TCP_KEEPIDLE`, `TCP_KEEPINTVL`, and `TCP_KEEPCNT` were not available on Solaris 
+     * until version 11.4, but let's take a chance here. */
+    #if defined(TCP_KEEPIDLE) && defined(TCP_KEEPINTVL) && defined(TCP_KEEPCNT)
+        if (setsockopt(fd, IPPROTO_TCP, TCP_KEEPIDLE, &idle, sizeof(idle))) {
+            anetSetError(err, "setsockopt TCP_KEEPIDLE: %s\n", strerror(errno));
+            return ANET_ERR;
+        }           
+        intvl = idle/3;
+        if (setsockopt(fd, IPPROTO_TCP, TCP_KEEPINTVL, &intvl, sizeof(intvl))) {
+            anetSetError(err, "setsockopt TCP_KEEPINTVL: %s\n", strerror(errno));
+            return ANET_ERR;
+        }
+        cnt = 3;
+        if (setsockopt(fd, IPPROTO_TCP, TCP_KEEPCNT, &cnt, sizeof(cnt))) {
+            anetSetError(err, "setsockopt TCP_KEEPCNT: %s\n", strerror(errno));
+            return ANET_ERR;
+        }
+        return ANET_OK;
+    #endif
+
+   /* Fall back to the first implementation of tcp-alive mechanism for older Solaris, 
+    * simulate the tcp-alive mechanism on other platforms via `TCP_KEEPALIVE_THRESHOLD` + `TCP_KEEPALIVE_ABORT_THRESHOLD`.
+    */
+    idle *= 1000; // kernel expects milliseconds
+    if (setsockopt(fd, IPPROTO_TCP, TCP_KEEPALIVE_THRESHOLD, &idle, sizeof(idle))) {
+        anetSetError(err, "setsockopt TCP_KEEPINTVL: %s\n", strerror(errno));
+        return ANET_ERR;
+    }
+
+    /* Note that the consequent probes will not be sent at equal intervals on Solaris, 
+     * but will be sent using the exponential backoff algorithm. */
+    intvl = idle/3;
+    cnt = 3;
+    int time_to_abort = intvl * cnt;
+    if (setsockopt(fd, IPPROTO_TCP, TCP_KEEPALIVE_ABORT_THRESHOLD, &time_to_abort, sizeof(time_to_abort))) {
+        anetSetError(err, "setsockopt TCP_KEEPCNT: %s\n", strerror(errno));
+        return ANET_ERR;
+    }
+
+    return ANET_OK;
+#endif
+
+#ifdef TCP_KEEPIDLE
```

Comment:
i'm curious, do we have a documented guarantee that this is a macro?
i.e. it could also be a `static const int TCP_KEEPIDLE = 4;` in some system, and this check would break.

Reply:
maybe we should remove the `UNUSED` lines, so that if on some system these are undefined, we'll at least get a compilation warning?

Reply:
> i'm curious, do we have a documented guarantee that this is a macro? i.e. it could also be a `static const int TCP_KEEPIDLE = 4;` in some system, and this check would break.

I think this shouldn't be a concern cuz `TCP_KEEPIDLE`, `TCP_KEEPINTVL`, and `TCP_KEEPCNT` are defined as macro in the kernel source code among those Unix-like OS's I referred to in the initial [comment](https://github.com/redis/redis/pull/12782/#issue-2001302294) of this PR, you can easily find these three macros used in some official C examples of those OS's, furthermore, you can find those macros in the kernel source code.

Reply:
are there `ifdef`s in the official examples?
or anything else that assumes this is a preprocessor definition?

Reply:
what would be the impact of removing the UNUSED lines?

Reply:
> are there `ifdef`s in the official examples? or anything else that assumes this is a preprocessor definition?

When the official example C code configure TCP keep-alive on specific OS, they don't need to use `#ifdef` because these macros are certainly defined in `<netinet/tcp.h>`, therefore they just use them by including that header file.

We can locate those macros in the kernel code:
- [freebsd](https://github.com/freebsd/freebsd-src/blob/5bc10feacc9d81e3bba9d28734a85e996682b408/sys/netinet/tcp.h#L204-L206)
- [dragonflybsd](https://github.com/DragonFlyBSD/DragonFlyBSD/blob/89ac0fe9e4b7cf2ffb93e881ef3b5a5ed8060a08/sys/netinet/tcp.h#L165-L167)
- [netbsd](https://github.com/NetBSD/src/blob/7426efd69687ed275130dc2eb761380fedda0398/sys/netinet/tcp.h#L127-L132)
- [illumos](https://github.com/illumos/illumos-gate/blob/a5f20e891eb1b5d2af2255e766a1dc3b35ac9936/usr/src/uts/common/netinet/tcp.h#L130-L132)

As for other non-open-sourced OS's like HP-UX, we can still find the answer from their [docs](https://docstore.mik.ua/manuals/hp-ux/en/B2355-60130/TCP.7P.html):
> The system supports the following socket options: TCP_MAXSEG, TCP_NODELAY, TCP_ABORT_THRESHOLD, TCP_CONN_ABORT_THRESHOLD, TCP_KEEPCNT, TCP_KEEPIDLE, TCP_KEEPINTVL, TCP_TSOPTENA, and TCP_SACKENA (defined in the include file ).

Reply:
> what would be the impact of removing the UNUSED lines?

I'd say it has no other impact other than maybe printing some warning outputs, I just don't see the need to remove those lines.

Reply:
> these macros are certainly defined in `<netinet/tcp.h>`, therefore they just use them by including that header file

that's my point. being "defined" doesn't mean it's defined in the preprocessor, there's no guarantee in the docs that an `#ifdef` would work. it could in theory be a `static const int`, and all the examples will still work, but the code in this PR will silently stop working (the ifdef-ed code will be skipped) and we won't know about it.

that's why i suggest to remove these UNUSED lines, so that if the system doesn't support them, we'll get a warning.
or maybe we can at least make these UNUSED lines conditional (not to exist on systems were we expect the keep-alive feature to be supported.

this would mean that we did `restructure the code and make it more generic to avoid multiple OS-specific macros`, but we still have OS-specific macros that validate that the generic code covers these OSes.
It would be easy to read / review each of these separately (i.e. the code that implements the logic will be free of OS concerns), and the code that validates the supported OSes are handled, will be free of any implementation logic.

Reply:
Based on my experience, it's unlikely for OS's to define `static const` instead of macros for something like socket options when those values are already defined as macros in the kernel, but I can't say it would never happen, as you said before, they could ***in theory*** be done that way somehow, to which I don't object.

Making those `UNUSED` lines conditional sounds feasible to me, but I have another idea which might make more sense:
```c
#if (!defined(TCP_KEEPIDLE) && !defined(TCP_KEEPALIVE)) || !defined(TCP_KEEPINTVL) || !defined(TCP_KEEPCNT)
    anetSetError(err, "anetKeepAlive: %s", "missing TCP keep-alive macros");
    return;
#endif
```
What do you think?

Reply:
but now we completely lack support for systems that don't have them?
the current version in unstable handles it only on some OSes.
the initial version of this PR handled them on many mode, but kept redis functional that's missing.
and now redis will not work (compile, but fail to run) on such systems.

i'd rather fail compilation, or even better, just emit a build warning.
i see two options for that warning or compilation error:
1. warn on any system doesn't support this.
2. warn specifically on the systems that we know should support it, if for some reason our ifdefs fail (e.g. specifically do that for Linux, Mac, etc).

Reply:
Sorry, I don't understand why you said “failed to run”?

Reply:
I think the latest commit won't cause redis to not work on those platforms that don't support the full mechanism of TCP keep-alive. But I now do realize that it contradicts with the following logic of solaris, I think emitting warnings on the platforms we expect to support keep-alive sounds feasible to me.